### PR TITLE
Monitor 2.0

### PR DESCRIPTION
**Pull request by mrz1836**

### CODE REVIEW

File: action_blockheader.go [hunk 1]
```diff
@@ -44,3 +44,30 @@ func (c *Client) RecordBlockHeader(ctx context.Context, hash string, bh bc.Block
...
 	return blockHeader, nil
 }
+
+// GetUnsyncedBlockHeaders get all unsynced block headers
+func (c *Client) GetUnsyncedBlockHeaders(ctx context.Context) ([]*BlockHeader, error) {
+
+	// Check for existing NewRelic transaction
+	ctx = c.GetOrStartTxn(ctx, "get_unsynced_blockheaders")
+
+	return GetUnsyncedBlockHeaders(ctx, c.DefaultModelOptions()...)
```
Comment (mrz1836) [inline action_blockheader.go:54]: Why is this a public function?

File: action_blockheader.go [hunk 2]
```diff
@@ -44,3 +44,30 @@ func (c *Client) RecordBlockHeader(ctx context.Context, hash string, bh bc.Block
...
+}
+
+// GetLastBlockHeader get last block header
+func (c *Client) GetLastBlockHeader(ctx context.Context) (*BlockHeader, error) {
+
+	// Check for existing NewRelic transaction
+	ctx = c.GetOrStartTxn(ctx, "get_last_blockheader")
+
+	return GetLastBlockHeader(ctx, c.DefaultModelOptions()...)
```
Comment (mrz1836) [inline action_blockheader.go:63]: Why is this a public function?

File: chainstate/processor.go [hunk 3]
```diff
@@ -56,46 +56,72 @@ func (p *BloomProcessor) Logger() Logger {
 	return p.logger
 }
 
-// GetHash get the hash of the current filter
+// GetHash get the hash of the current Filter
```
Comment (mrz1836) [inline chainstate/processor.go:59]: The word "Filter" vs "filter" in all the comments... looks like a refactor might have caused this

File: model_blockheaders.go [hunk 4]
```diff
@@ -137,25 +204,27 @@ func (m *BlockHeader) Migrate(client datastore.ClientInterface) error {
...
-					} else {
-						m.Client().Logger().Info(ctx, "Successfully imported all block headers into database")
-					}
+		ctx := context.Background()
+		// check whether we have block header 0, then we do not import
+		blockHeader0, err := getBlockHeaderByHeight(ctx, 0, m.Client().DefaultModelOptions()...)
+		if err != nil {
+			// stop execution if block headers import is not successful
+			// the block headers state can be messed up if they are not imported, or half imported
+			panic(err.Error())
```
Comment (mrz1836) [inline model_blockheaders.go:213]: I would avoid using panics....

File: model_blockheaders.go [hunk 5]
```diff
@@ -137,25 +204,27 @@ func (m *BlockHeader) Migrate(client datastore.ClientInterface) error {
...
+		} else {
+			if blockHeader0 == nil {
+				// import block headers in the background
+				m.Client().Logger().Info(ctx, "Importing block headers into database")
+				err = m.importBlockHeaders(ctx, client, blockHeadersFile)
+				if err != nil {
+					// stop execution if block headers import is not successful
+					// the block headers state can be messed up if they are not imported, or half imported
+					panic(err.Error())
```
Comment (mrz1836) [inline model_blockheaders.go:222]: I would avoid using panics....

File: monitor_event_handler.go [hunk 6]
```diff
@@ -47,9 +103,100 @@ func (h *MonitorEventHandler) OnConnect(_ *centrifuge.Client, e centrifuge.Conne
...
+						err = blockHeader.Save(ctx)
+						if err != nil {
+							h.logger.Error(h.ctx, err.Error())
+						}
+					}
+				}
+			}
+		}
+
+		time.Sleep(30 * time.Second)
```
Comment (mrz1836) [inline monitor_event_handler.go:172]: Wait 30 seconds? This should be a constant (in definitions or config)

**Review by mrz1836 — COMMENTED** [review]
See my comments

File: action_blockheader.go (hunk 2 above)
Comment (mrz1836) [reply action_blockheader.go:63]: This should be a private model function

File: monitor_event_handler.go (hunk 6 above)
Comment (mrz1836) [reply monitor_event_handler.go:172]: Anytime we have values like this, IMO they should live as constant or an option, but all in one place to make it easy to find and configure

File: action_blockheader.go (hunk 1 above)
Comment (icellan) [reply action_blockheader.go:55]: All actions are public functions, right?

File: chainstate/processor.go (hunk 3 above)
Comment (icellan) [reply chainstate/processor.go]: Fixed

**Pull request merged**